### ISSUE TITLE
Extend LSTM layer

### ISSUE THREAD

**Issue by AndreMaz** [tensorflow/tfjs#2755]
#### TensorFlow.js version
v1.5.2

#### Describe feature request
I'm trying to implement [Bahdanau's Attention](https://arxiv.org/abs/1409.0473). To do it I need to slightly modify the behavior of LSTM layer. However, looking at 
https://github.com/tensorflow/tfjs/blob/c114d6c608536adbc54bbb1b67c1a2293e85575c/tfjs-layers/src/exports_layers.ts#L1490
it seems that it's impossible and the only way to build a custom LSTM is by extending `RNN` class.  Am I right? Is there a reason for not exporting the LSTM class?

Also, after reading the paper, it seems that the modification has to be done at each time `step`. I've looked at the source code and the `step` is computed by an anonymous function. So even if I extend the `RNN` class I still won't be able to modify `step`'s behavior. 

https://github.com/tensorflow/tfjs/blob/c114d6c608536adbc54bbb1b67c1a2293e85575c/tfjs-layers/src/layers/recurrent.ts#L732-L740

I've also looked at [`keras/layers/recurrent.py`](https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/layers/recurrent.py#L666-L673) and at https://github.com/keras-team/keras/issues/3329 and it seems that in Keras it is possible to achieve what I pretend to do. However, I'm not familiar with tfjs's code base so I might be missing something. Is it possible to achieve the desired behavior? If so, how?

**Comment by rthadur** [comment]
Probably same answers applies to this please check [here](https://github.com/tensorflow/tfjs/issues/2721#issuecomment-586419657) and let us know ?

**Comment by AndreMaz** [comment]
Thanks for the response. In this case, can you provide a more detailed description or an example about what can be (and cannot be) done with layers? I didn't found much info about it at the site and the only example that I've found is [tfjs-examples/custom-layer](https://github.com/tensorflow/tfjs-examples/blob/master/custom-layer/custom_layer.js). However, it's too simple so it's difficult to see to what degree can we customize a layer. BTW what's the purpose of `invokeCallHook`?
It's present in the `custom-layer` example but it's not documented

https://github.com/tensorflow/tfjs-examples/blob/908ee32750ba750a14d15caeb53115e2d3dda2b3/custom-layer/custom_layer.js#L70

**Comment by AndreMaz** [comment]
For example, assuming that I want to create a custom layer that wraps a dense layer.
My `Custom` class would look like this:
```js
const tf = require("@tensorflow/tfjs-node");

class Custom extends tf.layers.Layer {
  constructor(config) {
    super(config);
    // My dense layer
    this.dense = tf.layers.dense({ units: 10, name: "denseLayer" });
  }

  build(inputShape) {
    this.customDense.build(inputShape);
  }
  computeOutputShape(inputShape) {
    return inputShape;
  }
  call(input) {
    // Do some stuff with the input data
    return this.dense.apply(input);
  }
  static get className() {
    return "Custom";
  }
}
tf.serialization.registerClass(Custom);
```
But how can I register `this.dense` variables in my `Custom` Layer to make them trainable? The `addWeight()` method wouldn't work in case. So if I run

```js
  const model = tf.sequential();
  model.add(new Custom({ name: "Custom_Layer", inputShape: [10] }));
  model.summary();
```
I would get the following info:
```bash
Layer (type)                 Output shape              Param #   
=================================================================
Custom_Layer (Custom)        [null,10]                 0         
=================================================================
Total params: 0
Trainable params: 0
Non-trainable params: 0
_________________________________________________________________
```
But using directly the `tf.layers.dense`
```js
  const model = tf.sequential();
  model.add(tf.layers.dense({ units: 10, inputShape: [10] }));
  model.summary();
```
I get this
```bash
Layer (type)                 Output shape              Param #   
=================================================================
dense_Dense1 (Dense)         [null,10]                 110       
=================================================================
Total params: 110
Trainable params: 110
Non-trainable params: 0
_________________________________________________________________
```
With this being said, I would really appreciate if there were a more advanced guide/tutorial explaining how to customize the layers.

**Comment by AndreMaz** [comment]
Regarding my last comment https://github.com/tensorflow/tfjs/issues/2755#issuecomment-588283952, I've started playing with Python version of TF and I've noticed that there is a possibility of [composing and nesting layers](https://www.tensorflow.org/tutorials/customization/custom_layers#models_composing_layers). It's awesome and it really simplifies the development process as I don't have to create custom layers from scratch. Is there any plans of adding layer composition to TF.js?
